drake/systems/plants/KinematicsCache.h

```diff
@@ -0,0 +1,184 @@
+  GradientVar<Scalar, TWIST_SIZE, Eigen::Dynamic, TWIST_SIZE, DrakeJoint::MAX_NUM_VELOCITIES> motion_subspace_in_body; // gradient w.r.t. q_i only
+  GradientVar<Scalar, TWIST_SIZE, Eigen::Dynamic, TWIST_SIZE, DrakeJoint::MAX_NUM_VELOCITIES> motion_subspace_in_world; // gradient w.r.t. q
+  GradientVar<Scalar, Eigen::Dynamic, Eigen::Dynamic, DrakeJoint::MAX_NUM_VELOCITIES, DrakeJoint::MAX_NUM_POSITIONS> qdot_to_v; // gradient w.r.t. q
+  GradientVar<Scalar, Eigen::Dynamic, Eigen::Dynamic, DrakeJoint::MAX_NUM_POSITIONS, DrakeJoint::MAX_NUM_VELOCITIES> v_to_qdot; // gradient w.r.t. q
+  GradientVar<Scalar, TWIST_SIZE, TWIST_SIZE> inertia_in_world;
+  GradientVar<Scalar, TWIST_SIZE, TWIST_SIZE> crb_in_world;
+
+  /*
+   * Configuration and velocity dependent
+   */
+  GradientVar<Scalar, TWIST_SIZE, 1> twist_in_world; // gradient w.r.t. q only; gradient w.r.t. v is motion_subspace_in_world
+  GradientVar<Scalar, TWIST_SIZE, 1> motion_subspace_in_body_dot_times_v; // gradient w.r.t. q_i and v_i only
+  GradientVar<Scalar, TWIST_SIZE, 1> motion_subspace_in_world_dot_times_v; // gradient w.r.t. q and v
+
+public:
+  KinematicsCacheElement(Eigen::DenseIndex num_positions_robot, Eigen::DenseIndex num_velocities_robot, int num_positions_joint, int num_velocities_joint, int gradient_order) :
+      dtransform_to_world_dq(HOMOGENEOUS_TRANSFORM_SIZE, num_positions_robot),
+      motion_subspace_in_body(TWIST_SIZE, num_velocities_joint, num_positions_joint, gradient_order),
+      motion_subspace_in_world(TWIST_SIZE, num_velocities_joint, num_positions_robot, gradient_order),
+      qdot_to_v(num_velocities_joint, num_positions_joint, num_positions_robot, gradient_order),
+      v_to_qdot(num_positions_joint, num_velocities_joint, num_positions_robot, gradient_order),
+      inertia_in_world(TWIST_SIZE, TWIST_SIZE, num_positions_robot, gradient_order),
+      crb_in_world(TWIST_SIZE, TWIST_SIZE, num_positions_robot, gradient_order),
+      twist_in_world(TWIST_SIZE, 1, num_positions_robot, gradient_order),
+      motion_subspace_in_body_dot_times_v(TWIST_SIZE, 1, num_positions_joint + num_velocities_joint, gradient_order),
+      motion_subspace_in_world_dot_times_v(TWIST_SIZE, 1, num_positions_robot + num_velocities_robot, gradient_order)
+  {
+    // empty
+  }
+};
+
+template <typename Scalar>
+class KinematicsCache
+{
+private:
+  std::unordered_map<RigidBody const *, KinematicsCacheElement<Scalar>> elements;
+  Eigen::Matrix<Scalar, Eigen::Dynamic, 1> q;
+  Eigen::Matrix<Scalar, Eigen::Dynamic, 1> v;
+  bool velocity_vector_valid;
+
+public:
+  int gradient_order;
+  bool position_kinematics_cached;
+  bool gradients_cached;
+  bool velocity_kinematics_cached;
+  bool jdotV_cached;
+  int cached_inertia_gradients_order;
+
+public:
+  KinematicsCache(const std::vector<std::shared_ptr<RigidBody>>& bodies, int gradient_order) :
```

Comment:
what are you referring to? gradient_order since it only has two options? If so, gradient_order (here and everywhere in our codebase) will be removed when we switch to AutoDiffScalar based gradients.